// @buf/meshtastic_protobufs.bufbuild_es/meshtastic/module_config_pb.js
// @generated by protoc-gen-es v1.3.1
// @generated from file meshtastic/module_config.proto (package meshtastic, syntax proto3)
/* eslint-disable */
// @ts-nocheck

import { proto3 } from "@bufbuild/protobuf";

/**
 * @generated from enum meshtastic.RemoteHardwarePinType
 */
export const RemoteHardwarePinType = proto3.makeEnum(
  "meshtastic.RemoteHardwarePinType",
  [
    {no: 0, name: "UNKNOWN"},
    {no: 1, name: "DIGITAL_READ"},
    {no: 2, name: "DIGITAL_WRITE"},
  ],
);

/**
 *
 * Module Config
 *
 * @generated from message meshtastic.ModuleConfig
 */
export const ModuleConfig = proto3.makeMessageType(
  "meshtastic.ModuleConfig",
  () => [
    { no: 1, name: "mqtt", kind: "message", T: ModuleConfig_MQTTConfig, oneof: "payload_variant" },
    { no: 2, name: "serial", kind: "message", T: ModuleConfig_SerialConfig, oneof: "payload_variant" },
    { no: 3, name: "external_notification", kind: "message", T: ModuleConfig_ExternalNotificationConfig, oneof: "payload_variant" },
    { no: 4, name: "store_forward", kind: "message", T: ModuleConfig_StoreForwardConfig, oneof: "payload_variant" },
    { no: 5, name: "range_test", kind: "message", T: ModuleConfig_RangeTestConfig, oneof: "payload_variant" },
    { no: 6, name: "telemetry", kind: "message", T: ModuleConfig_TelemetryConfig, oneof: "payload_variant" },
    { no: 7, name: "canned_message", kind: "message", T: ModuleConfig_CannedMessageConfig, oneof: "payload_variant" },
    { no: 8, name: "audio", kind: "message", T: ModuleConfig_AudioConfig, oneof: "payload_variant" },
    { no: 9, name: "remote_hardware", kind: "message", T: ModuleConfig_RemoteHardwareConfig, oneof: "payload_variant" },
    { no: 10, name: "neighbor_info", kind: "message", T: ModuleConfig_NeighborInfoConfig, oneof: "payload_variant" },
    { no: 11, name: "ambient_lighting", kind: "message", T: ModuleConfig_AmbientLightingConfig, oneof: "payload_variant" },
    { no: 12, name: "detection_sensor", kind: "message", T: ModuleConfig_DetectionSensorConfig, oneof: "payload_variant" },
  ],
);

/**
 *
 * MQTT Client Config
 *
 * @generated from message meshtastic.ModuleConfig.MQTTConfig
 */
export const ModuleConfig_MQTTConfig = proto3.makeMessageType(
  "meshtastic.ModuleConfig.MQTTConfig",
  () => [
    { no: 1, name: "enabled", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 2, name: "address", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "username", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 4, name: "password", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 5, name: "encryption_enabled", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 6, name: "json_enabled", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 7, name: "tls_enabled", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 8, name: "root", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 9, name: "proxy_to_client_enabled", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
  ],
  {localName: "ModuleConfig_MQTTConfig"},
);

/**
 *
 * RemoteHardwareModule Config
 *
 * @generated from message meshtastic.ModuleConfig.RemoteHardwareConfig
 */
export const ModuleConfig_RemoteHardwareConfig = proto3.makeMessageType(
  "meshtastic.ModuleConfig.RemoteHardwareConfig",
  () => [
    { no: 1, name: "enabled", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 2, name: "allow_undefined_pin_access", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 3, name: "available_pins", kind: "message", T: RemoteHardwarePin, repeated: true },
  ],
  {localName: "ModuleConfig_RemoteHardwareConfig"},
);

/**
 *
 * NeighborInfoModule Config
 *
 * @generated from message meshtastic.ModuleConfig.NeighborInfoConfig
 */
export const ModuleConfig_NeighborInfoConfig = proto3.makeMessageType(
  "meshtastic.ModuleConfig.NeighborInfoConfig",
  () => [
    { no: 1, name: "enabled", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 2, name: "update_interval", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
  ],
  {localName: "ModuleConfig_NeighborInfoConfig"},
);

/**
 *
 * Detection Sensor Module Config
 *
 * @generated from message meshtastic.ModuleConfig.DetectionSensorConfig
 */
export const ModuleConfig_DetectionSensorConfig = proto3.makeMessageType(
  "meshtastic.ModuleConfig.DetectionSensorConfig",
  () => [
    { no: 1, name: "enabled", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 2, name: "minimum_broadcast_secs", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 3, name: "state_broadcast_secs", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 4, name: "send_bell", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 5, name: "name", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 6, name: "monitor_pin", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 7, name: "detection_triggered_high", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 8, name: "use_pullup", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
  ],
  {localName: "ModuleConfig_DetectionSensorConfig"},
);

/**
 *
 * Audio Config for codec2 voice
 *
 * @generated from message meshtastic.ModuleConfig.AudioConfig
 */
export const ModuleConfig_AudioConfig = proto3.makeMessageType(
  "meshtastic.ModuleConfig.AudioConfig",
  () => [
    { no: 1, name: "codec2_enabled", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 2, name: "ptt_pin", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 3, name: "bitrate", kind: "enum", T: proto3.getEnumType(ModuleConfig_AudioConfig_Audio_Baud) },
    { no: 4, name: "i2s_ws", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 5, name: "i2s_sd", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 6, name: "i2s_din", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 7, name: "i2s_sck", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
  ],
  {localName: "ModuleConfig_AudioConfig"},
);

/**
 *
 * Baudrate for codec2 voice
 *
 * @generated from enum meshtastic.ModuleConfig.AudioConfig.Audio_Baud
 */
export const ModuleConfig_AudioConfig_Audio_Baud = proto3.makeEnum(
  "meshtastic.ModuleConfig.AudioConfig.Audio_Baud",
  [
    {no: 0, name: "CODEC2_DEFAULT"},
    {no: 1, name: "CODEC2_3200"},
    {no: 2, name: "CODEC2_2400"},
    {no: 3, name: "CODEC2_1600"},
    {no: 4, name: "CODEC2_1400"},
    {no: 5, name: "CODEC2_1300"},
    {no: 6, name: "CODEC2_1200"},
    {no: 7, name: "CODEC2_700"},
    {no: 8, name: "CODEC2_700B"},
  ],
);

/**
 *
 * Serial Config
 *
 * @generated from message meshtastic.ModuleConfig.SerialConfig
 */
export const ModuleConfig_SerialConfig = proto3.makeMessageType(
  "meshtastic.ModuleConfig.SerialConfig",
  () => [
    { no: 1, name: "enabled", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 2, name: "echo", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 3, name: "rxd", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 4, name: "txd", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 5, name: "baud", kind: "enum", T: proto3.getEnumType(ModuleConfig_SerialConfig_Serial_Baud) },
    { no: 6, name: "timeout", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 7, name: "mode", kind: "enum", T: proto3.getEnumType(ModuleConfig_SerialConfig_Serial_Mode) },
    { no: 8, name: "override_console_serial_port", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
  ],
  {localName: "ModuleConfig_SerialConfig"},
);

/**
 *
 * TODO: REPLACE
 *
 * @generated from enum meshtastic.ModuleConfig.SerialConfig.Serial_Baud
 */
export const ModuleConfig_SerialConfig_Serial_Baud = proto3.makeEnum(
  "meshtastic.ModuleConfig.SerialConfig.Serial_Baud",
  [
    {no: 0, name: "BAUD_DEFAULT"},
    {no: 1, name: "BAUD_110"},
    {no: 2, name: "BAUD_300"},
    {no: 3, name: "BAUD_600"},
    {no: 4, name: "BAUD_1200"},
    {no: 5, name: "BAUD_2400"},
    {no: 6, name: "BAUD_4800"},
    {no: 7, name: "BAUD_9600"},
    {no: 8, name: "BAUD_19200"},
    {no: 9, name: "BAUD_38400"},
    {no: 10, name: "BAUD_57600"},
    {no: 11, name: "BAUD_115200"},
    {no: 12, name: "BAUD_230400"},
    {no: 13, name: "BAUD_460800"},
    {no: 14, name: "BAUD_576000"},
    {no: 15, name: "BAUD_921600"},
  ],
);

/**
 *
 * TODO: REPLACE
 *
 * @generated from enum meshtastic.ModuleConfig.SerialConfig.Serial_Mode
 */
export const ModuleConfig_SerialConfig_Serial_Mode = proto3.makeEnum(
  "meshtastic.ModuleConfig.SerialConfig.Serial_Mode",
  [
    {no: 0, name: "DEFAULT"},
    {no: 1, name: "SIMPLE"},
    {no: 2, name: "PROTO"},
    {no: 3, name: "TEXTMSG"},
    {no: 4, name: "NMEA"},
    {no: 5, name: "CALTOPO"},
  ],
);

/**
 *
 * External Notifications Config
 *
 * @generated from message meshtastic.ModuleConfig.ExternalNotificationConfig
 */
export const ModuleConfig_ExternalNotificationConfig = proto3.makeMessageType(
  "meshtastic.ModuleConfig.ExternalNotificationConfig",
  () => [
    { no: 1, name: "enabled", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 2, name: "output_ms", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 3, name: "output", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 8, name: "output_vibra", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 9, name: "output_buzzer", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 4, name: "active", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 5, name: "alert_message", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 10, name: "alert_message_vibra", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 11, name: "alert_message_buzzer", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 6, name: "alert_bell", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 12, name: "alert_bell_vibra", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 13, name: "alert_bell_buzzer", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 7, name: "use_pwm", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 14, name: "nag_timeout", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
  ],
  {localName: "ModuleConfig_ExternalNotificationConfig"},
);

/**
 *
 * Store and Forward Module Config
 *
 * @generated from message meshtastic.ModuleConfig.StoreForwardConfig
 */
export const ModuleConfig_StoreForwardConfig = proto3.makeMessageType(
  "meshtastic.ModuleConfig.StoreForwardConfig",
  () => [
    { no: 1, name: "enabled", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 2, name: "heartbeat", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 3, name: "records", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 4, name: "history_return_max", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 5, name: "history_return_window", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
  ],
  {localName: "ModuleConfig_StoreForwardConfig"},
);

/**
 *
 * Preferences for the RangeTestModule
 *
 * @generated from message meshtastic.ModuleConfig.RangeTestConfig
 */
export const ModuleConfig_RangeTestConfig = proto3.makeMessageType(
  "meshtastic.ModuleConfig.RangeTestConfig",
  () => [
    { no: 1, name: "enabled", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 2, name: "sender", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 3, name: "save", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
  ],
  {localName: "ModuleConfig_RangeTestConfig"},
);

/**
 *
 * Configuration for both device and environment metrics
 *
 * @generated from message meshtastic.ModuleConfig.TelemetryConfig
 */
export const ModuleConfig_TelemetryConfig = proto3.makeMessageType(
  "meshtastic.ModuleConfig.TelemetryConfig",
  () => [
    { no: 1, name: "device_update_interval", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 2, name: "environment_update_interval", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 3, name: "environment_measurement_enabled", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 4, name: "environment_screen_enabled", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 5, name: "environment_display_fahrenheit", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 6, name: "air_quality_enabled", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 7, name: "air_quality_interval", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
  ],
  {localName: "ModuleConfig_TelemetryConfig"},
);

/**
 *
 * TODO: REPLACE
 *
 * @generated from message meshtastic.ModuleConfig.CannedMessageConfig
 */
export const ModuleConfig_CannedMessageConfig = proto3.makeMessageType(
  "meshtastic.ModuleConfig.CannedMessageConfig",
  () => [
    { no: 1, name: "rotary1_enabled", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 2, name: "inputbroker_pin_a", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 3, name: "inputbroker_pin_b", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 4, name: "inputbroker_pin_press", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 5, name: "inputbroker_event_cw", kind: "enum", T: proto3.getEnumType(ModuleConfig_CannedMessageConfig_InputEventChar) },
    { no: 6, name: "inputbroker_event_ccw", kind: "enum", T: proto3.getEnumType(ModuleConfig_CannedMessageConfig_InputEventChar) },
    { no: 7, name: "inputbroker_event_press", kind: "enum", T: proto3.getEnumType(ModuleConfig_CannedMessageConfig_InputEventChar) },
    { no: 8, name: "updown1_enabled", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 9, name: "enabled", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 10, name: "allow_input_source", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 11, name: "send_bell", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
  ],
  {localName: "ModuleConfig_CannedMessageConfig"},
);

/**
 *
 * TODO: REPLACE
 *
 * @generated from enum meshtastic.ModuleConfig.CannedMessageConfig.InputEventChar
 */
export const ModuleConfig_CannedMessageConfig_InputEventChar = proto3.makeEnum(
  "meshtastic.ModuleConfig.CannedMessageConfig.InputEventChar",
  [
    {no: 0, name: "NONE"},
    {no: 17, name: "UP"},
    {no: 18, name: "DOWN"},
    {no: 19, name: "LEFT"},
    {no: 20, name: "RIGHT"},
    {no: 10, name: "SELECT"},
    {no: 27, name: "BACK"},
    {no: 24, name: "CANCEL"},
  ],
);

/**
 * Ambient Lighting Module - Settings for control of onboard LEDs to allow users to adjust the brightness levels and respective color levels.
 * Initially created for the RAK14001 RGB LED module.
 *
 * @generated from message meshtastic.ModuleConfig.AmbientLightingConfig
 */
export const ModuleConfig_AmbientLightingConfig = proto3.makeMessageType(
  "meshtastic.ModuleConfig.AmbientLightingConfig",
  () => [
    { no: 1, name: "led_state", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 2, name: "current", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 3, name: "red", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 4, name: "green", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 5, name: "blue", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
  ],
  {localName: "ModuleConfig_AmbientLightingConfig"},
);

/**
 *
 * A GPIO pin definition for remote hardware module
 *
 * @generated from message meshtastic.RemoteHardwarePin
 */
export const RemoteHardwarePin = proto3.makeMessageType(
  "meshtastic.RemoteHardwarePin",
  () => [
    { no: 1, name: "gpio_pin", kind: "scalar", T: 13 /* ScalarType.UINT32 */ },
    { no: 2, name: "name", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "type", kind: "enum", T: proto3.getEnumType(RemoteHardwarePinType) },
  ],
);
