// @meshtastic/meshtasticjs/dist/utils/queue.js
import { Protobuf } from "../index.js";
import { SubEvent } from "sub-events";
export class Queue {
    queue = [];
    lock = false;
    ackNotifier = new SubEvent();
    errorNotifier = new SubEvent();
    timeout;
    constructor() {
        this.timeout = 60000;
    }
    getState() {
        return this.queue;
    }
    clear() {
        this.queue = [];
    }
    push(item) {
        const queueItem = {
            ...item,
            sent: false,
            added: new Date(),
            promise: new Promise((resolve, reject) => {
                this.ackNotifier.subscribe((id) => {
                    if (item.id === id) {
                        this.remove(item.id);
                        resolve(id);
                    }
                });
                this.errorNotifier.subscribe((e) => {
                    if (item.id === e.id) {
                        this.remove(item.id);
                        reject(e);
                    }
                });
                setTimeout(() => {
                    if (this.queue.findIndex((qi) => qi.id === item.id) !== -1) {
                        this.remove(item.id);
                        const decoded = Protobuf.ToRadio.fromBinary(item.data);
                        console.warn(`Packet ${item.id} of type ${decoded.payloadVariant.case} timed out`);
                        reject({
                            id: item.id,
                            error: Protobuf.Routing_Error.TIMEOUT,
                        });
                    }
                }, this.timeout);
            }),
        };
        this.queue.push(queueItem);
    }
    remove(id) {
        if (this.lock) {
            setTimeout(() => this.remove(id), 100);
            return;
        }
        this.queue = this.queue.filter((item) => item.id !== id);
    }
    processAck(id) {
        this.ackNotifier.emit(id);
    }
    processError(e) {
        console.error(`Error received for packet ${e.id}: ${Protobuf.Routing_Error[e.error]}`);
        this.errorNotifier.emit(e);
    }
    async wait(id) {
        const queueItem = this.queue.find((qi) => qi.id === id);
        if (!queueItem) {
            throw new Error("Packet does not exist");
        }
        return queueItem.promise;
    }
    async processQueue(writeToRadio) {
        if (this.lock) {
            return;
        }
        this.lock = true;
        while (this.queue.filter((p) => !p.sent).length > 0) {
            const item = this.queue.filter((p) => !p.sent)[0];
            if (item) {
                await new Promise((resolve) => setTimeout(resolve, 200));
                try {
                    await writeToRadio(item.data);
                    item.sent = true;
                }
                catch (error) {
                    console.error(`Error sending packet ${item.id}`, error);
                }
            }
        }
        this.lock = false;
    }
}
