// @meshtastic/meshtasticjs/dist/types.js
export var DeviceStatusEnum;
(function (DeviceStatusEnum) {
    DeviceStatusEnum[DeviceStatusEnum["DEVICE_RESTARTING"] = 1] = "DEVICE_RESTARTING";
    DeviceStatusEnum[DeviceStatusEnum["DEVICE_DISCONNECTED"] = 2] = "DEVICE_DISCONNECTED";
    DeviceStatusEnum[DeviceStatusEnum["DEVICE_CONNECTING"] = 3] = "DEVICE_CONNECTING";
    DeviceStatusEnum[DeviceStatusEnum["DEVICE_RECONNECTING"] = 4] = "DEVICE_RECONNECTING";
    DeviceStatusEnum[DeviceStatusEnum["DEVICE_CONNECTED"] = 5] = "DEVICE_CONNECTED";
    DeviceStatusEnum[DeviceStatusEnum["DEVICE_CONFIGURING"] = 6] = "DEVICE_CONFIGURING";
    DeviceStatusEnum[DeviceStatusEnum["DEVICE_CONFIGURED"] = 7] = "DEVICE_CONFIGURED";
})(DeviceStatusEnum || (DeviceStatusEnum = {}));
export var EmitterScope;
(function (EmitterScope) {
    EmitterScope[EmitterScope["iMeshDevice"] = 1] = "iMeshDevice";
    EmitterScope[EmitterScope["iSerialConnection"] = 2] = "iSerialConnection";
    EmitterScope[EmitterScope["iNodeSerialConnection"] = 3] = "iNodeSerialConnection";
    EmitterScope[EmitterScope["iBleConnection"] = 4] = "iBleConnection";
    EmitterScope[EmitterScope["iHttpConnection"] = 5] = "iHttpConnection";
})(EmitterScope || (EmitterScope = {}));
export var Emitter;
(function (Emitter) {
    Emitter[Emitter["constructor"] = 0] = "constructor";
    Emitter[Emitter["sendText"] = 1] = "sendText";
    Emitter[Emitter["sendWaypoint"] = 2] = "sendWaypoint";
    Emitter[Emitter["sendPacket"] = 3] = "sendPacket";
    Emitter[Emitter["sendRaw"] = 4] = "sendRaw";
    Emitter[Emitter["setConfig"] = 5] = "setConfig";
    Emitter[Emitter["setModuleConfig"] = 6] = "setModuleConfig";
    Emitter[Emitter["confirmSetConfig"] = 7] = "confirmSetConfig";
    Emitter[Emitter["setOwner"] = 8] = "setOwner";
    Emitter[Emitter["setChannel"] = 9] = "setChannel";
    Emitter[Emitter["confirmSetChannel"] = 10] = "confirmSetChannel";
    Emitter[Emitter["clearChannel"] = 11] = "clearChannel";
    Emitter[Emitter["getChannel"] = 12] = "getChannel";
    Emitter[Emitter["getAllChannels"] = 13] = "getAllChannels";
    Emitter[Emitter["getConfig"] = 14] = "getConfig";
    Emitter[Emitter["getModuleConfig"] = 15] = "getModuleConfig";
    Emitter[Emitter["getOwner"] = 16] = "getOwner";
    Emitter[Emitter["configure"] = 17] = "configure";
    Emitter[Emitter["handleFromRadio"] = 18] = "handleFromRadio";
    Emitter[Emitter["handleMeshPacket"] = 19] = "handleMeshPacket";
    Emitter[Emitter["connect"] = 20] = "connect";
    Emitter[Emitter["ping"] = 21] = "ping";
    Emitter[Emitter["readFromRadio"] = 22] = "readFromRadio";
    Emitter[Emitter["writeToRadio"] = 23] = "writeToRadio";
    Emitter[Emitter["setDebugMode"] = 24] = "setDebugMode";
    Emitter[Emitter["getMetadata"] = 25] = "getMetadata";
    Emitter[Emitter["resetPeers"] = 26] = "resetPeers";
    Emitter[Emitter["shutdown"] = 27] = "shutdown";
    Emitter[Emitter["reboot"] = 28] = "reboot";
    Emitter[Emitter["rebootOTA"] = 29] = "rebootOTA";
    Emitter[Emitter["factoryReset"] = 30] = "factoryReset";
})(Emitter || (Emitter = {}));
export var ChannelNumber;
(function (ChannelNumber) {
    ChannelNumber[ChannelNumber["PRIMARY"] = 0] = "PRIMARY";
    ChannelNumber[ChannelNumber["CHANNEL1"] = 1] = "CHANNEL1";
    ChannelNumber[ChannelNumber["CHANNEL2"] = 2] = "CHANNEL2";
    ChannelNumber[ChannelNumber["CHANNEL3"] = 3] = "CHANNEL3";
    ChannelNumber[ChannelNumber["CHANNEL4"] = 4] = "CHANNEL4";
    ChannelNumber[ChannelNumber["CHANNEL5"] = 5] = "CHANNEL5";
    ChannelNumber[ChannelNumber["CHANNEL6"] = 6] = "CHANNEL6";
    ChannelNumber[ChannelNumber["ADMIN"] = 7] = "ADMIN";
})(ChannelNumber || (ChannelNumber = {}));
