// @meshtastic/meshtasticjs/dist/adapters/iBleConnection.js
import { fromNumUUID, fromRadioUUID, serviceUUID, toRadioUUID, } from "../constants.js";
import { IMeshDevice } from "../iMeshDevice.js";
import { Types } from "../index.js";
import { typedArrayToBuffer } from "../utils/general.js";
/** Allows to connect to a Meshtastic device via bluetooth */
export class IBLEConnection extends IMeshDevice {
    /** Defines the connection type as ble */
    connType;
    portId;
    /** Currently connected BLE device */
    device;
    GATTServer;
    /** Short Description */
    service;
    /** Short Description */
    toRadioCharacteristic;
    /** Short Description */
    fromRadioCharacteristic;
    /** Short Description */
    fromNumCharacteristic;
    constructor(configId) {
        super(configId);
        this.log = this.log.getSubLogger({ name: "iHttpConnection" });
        this.connType = "ble";
        this.portId = "";
        this.device = undefined;
        this.service = undefined;
        this.GATTServer = undefined;
        this.toRadioCharacteristic = undefined;
        this.fromRadioCharacteristic = undefined;
        this.fromNumCharacteristic = undefined;
        // this.pendingRead = false;
        this.log.debug(Types.Emitter[Types.Emitter.constructor], "🔷 iBleConnection instantiated");
    }
    /**
     * Gets web bluetooth support avaliability for the device
     *
     * @returns {Promise<void>}
     */
    supported() {
        return navigator.bluetooth.getAvailability();
    }
    /**
     * Gets list of bluetooth devices that can be passed to `connect`
     *
     * @returns {Promise<BluetoothDevice[]>} Array of avaliable BLE devices
     */
    getDevices() {
        return navigator.bluetooth.getDevices();
    }
    /**
     * Opens browser dialog to select a device
     */
    getDevice(filter) {
        return navigator.bluetooth.requestDevice(filter ?? {
            filters: [{ services: [serviceUUID] }],
        });
    }
    /**
     * Initiates the connect process to a Meshtastic device via Bluetooth
     */
    async connect({ device, deviceFilter, }) {
        /** Set device state to connecting */
        this.updateDeviceStatus(Types.DeviceStatusEnum.DEVICE_CONNECTING);
        /** Set device if specified, else request. */
        this.device = device ?? (await this.getDevice(deviceFilter));
        this.portId = this.device.id;
        /** Setup event listners */
        this.device.addEventListener("gattserverdisconnected", () => {
            this.log.info(Types.Emitter[Types.Emitter.connect], "Device disconnected");
            this.updateDeviceStatus(Types.DeviceStatusEnum.DEVICE_DISCONNECTED);
            this.complete();
        });
        /** Connect to device */
        await this.device.gatt
            ?.connect()
            .then((server) => {
            this.log.info(Types.Emitter[Types.Emitter.connect], `✅ Got GATT Server for device: ${server.device.id}`);
            this.GATTServer = server;
        })
            .catch((e) => {
            this.log.error(Types.Emitter[Types.Emitter.connect], `❌ Failed to connect: ${e.message}`);
        });
        await this.GATTServer?.getPrimaryService(serviceUUID)
            .then((service) => {
            this.log.info(Types.Emitter[Types.Emitter.connect], `✅ Got GATT Service for device: ${service.device.id}`);
            this.service = service;
        })
            .catch((e) => {
            this.log.error(Types.Emitter[Types.Emitter.connect], `❌ Failed to get primary service: q${e.message}`);
        });
        [toRadioUUID, fromRadioUUID, fromNumUUID].map(async (uuid) => {
            await this.service
                ?.getCharacteristic(uuid)
                .then((characteristic) => {
                this.log.info(Types.Emitter[Types.Emitter.connect], `✅ Got Characteristic ${characteristic.uuid} for device: ${characteristic.uuid}`);
                switch (uuid) {
                    case toRadioUUID:
                        this.toRadioCharacteristic = characteristic;
                        break;
                    case fromRadioUUID:
                        this.fromRadioCharacteristic = characteristic;
                        break;
                    case fromNumUUID:
                        this.fromNumCharacteristic = characteristic;
                        break;
                }
            })
                .catch((e) => {
                this.log.error(Types.Emitter[Types.Emitter.connect], `❌ Failed to get toRadio characteristic: q${e.message}`);
            });
        });
        await this.fromNumCharacteristic?.startNotifications(); // TODO: catch
        this.fromNumCharacteristic?.addEventListener("characteristicvaluechanged", () => {
            void this.readFromRadio();
        });
        this.updateDeviceStatus(Types.DeviceStatusEnum.DEVICE_CONNECTED);
        void this.configure().catch(() => {
            // TODO: FIX, workaround for `wantConfigId` not getting acks.
        });
    }
    /** Disconnects from the Meshtastic device */
    disconnect() {
        this.device?.gatt?.disconnect();
        this.updateDeviceStatus(Types.DeviceStatusEnum.DEVICE_DISCONNECTED);
        this.complete();
    }
    /**
     * Pings device to check if it is avaliable
     *
     * @todo Implement
     */
    async ping() {
        return Promise.resolve(true);
    }
    /** Short description */
    async readFromRadio() {
        // if (this.pendingRead) {
        //   return Promise.resolve();
        // }
        // this.pendingRead = true;
        let readBuffer = new ArrayBuffer(1);
        while (readBuffer.byteLength > 0 && this.fromRadioCharacteristic) {
            await this.fromRadioCharacteristic
                .readValue()
                .then((value) => {
                readBuffer = value.buffer;
                if (value.byteLength > 0) {
                    this.handleFromRadio(new Uint8Array(readBuffer));
                }
                this.updateDeviceStatus(Types.DeviceStatusEnum.DEVICE_CONNECTED);
            })
                .catch((e) => {
                readBuffer = new ArrayBuffer(0);
                this.log.error(Types.Emitter[Types.Emitter.readFromRadio], `❌ ${e.message}`);
            });
        }
        // this.pendingRead = false;
    }
    /**
     * Sends supplied protobuf message to the radio
     */
    async writeToRadio(data) {
        await this.toRadioCharacteristic?.writeValue(typedArrayToBuffer(data));
        // This should be automatic (onCharacteristicValueChanged)
        await this.readFromRadio();
    }
}
